Composable auspice config

jameshadfield opened the pull request:
Closes #298 

See commit messages for details, but there's a few parts I want to call out explicitly:

### Merging approach

Specifically, merging of lists is done by extending the original list, rather than simply replacing it. We use a custom approach to decide whether new elements should be added to the list or replace already-present elements. As an example:

```json
{
  "colorings": [
    {"key": "num_date", "title": "date", "type": "continuous"}
  ]
}
```

```json
{
  "colorings": [
    {"key": "country", "title": "Country", "type": "categorical"},
    {"key": "num_date", "title": "Sampling Date", "type": "continuous"}
  ]
}
```

Here the second JSON's `num_date` coloring replaces the first JSONs one, and the `country` coloring is added in, so we end up with:

```json
{
  "colorings": [
    {"key": "num_date", "title": "Sampling Date", "type": "continuous"}
    {"key": "country", "title": "Country", "type": "categorical"},
  ]
}
```

I think this is the only practical way to combine colorings in multiple configs, but we also use lists for many other options, e.g. 'panels', 'maintainers'. This PR still uses the same approach, but I'm less convinced for non-coloring options such as these whether we should extend lists or simply replace them.

Note that we don't really do deep merging here at all. This only applies in one situation as far as I can tell, [`config.display_defaults.panels: list[str]`](https://github.com/nextstrain/augur/blob/bbadca3cd42e0de67016a951759bad5579aa96f5/augur/data/schema-auspice-config-v2.json#L211-L219) where overlay configs will completely replace the original list, and this is different from how `config.panels` currently behaves.


### Refactoring of warnings code

This changed all usages of (export v2's) `warn()` to `warning()`, which in practice only changes [the stacklevel](https://docs.python.org/3/library/warnings.html#warnings.warn), which changes... nothing? The `warning()` helper function was added [years ago](https://github.com/nextstrain/augur/commit/b225bd00fe71330ae93f846e3991bdc66f434000) but saw ~no usage, so I think we should either use it consistently (as done here) or remove it and just use the default `warn()`. 

### Validate then merge

The approach here is to validate each config JSON independently, rather than merging then validating. (This works well since our schema has no required properties.) `augur validate auspice-config-v2` _hasn't_ been updated to allow multiple auspice configs, but we could do so if we want (requires a little refactoring to avoid circular imports). We could additionally validate the merged config if people want to, although the current implementation guarantees that valid configs produce a valid merged config.

### Long term direction

This work paves the way to merging in all config-related command line args within `auspice_config.py` which should help clean up `export_v2.py` quite a bit.

I didn't add an option to write out the merged config, but I know this is a subject that's been discussed quite a bit in the context of snakemake config merging so we could add this here if wanted.


### Still to do:
- [x] regenerate docs. I need to follow Jover's instructions here as with my setup (py 3.10) I introduce changes to every docs file.
- [x] changelog

victorlin reviewed (commented):
Only reviewed the first 2 commits so far, sending in comments before lab meeting

victorlin commented on augur/export_v2.py:
```diff
@@ -264,7 +211,7 @@ def _get_type(key, trait_values):
             allowedTypes = ["continuous", "temporal", "ordinal", "categorical", "boolean"]
             if t not in allowedTypes:
-                warn("[config file] In trait {}, coloring type '{}' is not valid. Please choose from: '{}'. This trait has been excluded!".format(key, t, ", ".join(allowedTypes)))
+                warning("[config file] In trait {}, coloring type '{}' is not valid. Please choose from: '{}'. This trait has been excluded!".format(key, t, ", ".join(allowedTypes)))
```
No, I think the wrapper function is a good change – my suggestion is to rename it to `warn`.

jameshadfield replied:
>  I think the wrapper function is a good change – my suggestion is to rename it to warn.

Gotcha. Done in 998e2a1ea17dc6a497e7070857c15b3f164bc220

victorlin commented on augur/util_support/warnings.py:
```diff
@@ -0,0 +1,34 @@
+def warning(message):
+    warn(message, UserWarning, stacklevel=2)
+
+def fatal(message):
+    print("FATAL ERROR: {}".format(message))
+    sys.exit(2)
```
Fair, consider this non-blocking!

jameshadfield replied:
Fixed via stand-alone commit 3492d3c666f0d601b5adc926c7539b4a7100349b. We actually use `fatal`  twice (within `export_v2.py`) and call `AugurError` twice so I opted to move all usages to the latter. As you note, this changes the printed message slightly, but it's now more consistent with other code in the project.

genehack commented on augur/util_support/auspice_config.py:
```diff
@@ -0,0 +1,234 @@
+        raise AugurError(f"[config file] The deprecated field 'color_options' must be a dictionary")
+    colorings = []
+    for key, info in options.items():
+        # note: if both legentTitle & menuItem are present then we use the latter. See https://github.com/nextstrain/auspice/issues/730
```
```suggestion
        # note: if both legendTitle & menuItem are present then we use the latter. See https://github.com/nextstrain/auspice/issues/730
```

jameshadfield replied:
Good eyes - this typo wasn't introduced by me (it exists in current master) but I've fixed it up

genehack commented on augur/util_support/auspice_config.py:
```diff
@@ -0,0 +1,234 @@
+                deprecated("[config file] coloring '{}': 'legendTitle' has been replaced with 'title'. Using 'legendTitle' as 'title'.".format(key))
+                info["title"] = info["legendTitle"]
+            del info["legendTitle"]
+        if "key" in info:
+            del info["key"]
```
No deprecation message here?

jameshadfield replied:
Copied as is from [current master code](https://github.com/nextstrain/augur/blob/04d1a00e84a7cc5f5209a3d022de1114e92772f2/augur/export_v2.py#L240-L241). I tried not to dwell too much about improving deprecation messaging - I think we should be more proactive about dropping support if anything!

genehack commented on augur/util_support/auspice_config.py:
```diff
@@ -0,0 +1,234 @@
+        if "key" in info:
+            del info["key"]
+        if info.get("type") == "discrete":
+            deprecated("[config file] coloring '{}': type 'discrete' is no longer valid. Please use either 'ordinal', 'categorical' or 'boolean'. "
```
```suggestion
            deprecated("[config file] coloring '{}': type 'discrete' is no longer valid. Please use 'ordinal', 'categorical' or 'boolean'. "
```

jameshadfield replied:
This wasn't introduced by me (it exists in current master) but your suggestion is better & we're now using it

genehack commented on augur/util_support/auspice_config.py:
```diff
@@ -0,0 +1,234 @@
+    authors = next(iter([{'index': idx, 'coloring': c} for [idx, c] in enumerate(colorings) if c['key']=='authors'] or [None]))
+    if authors:
+        if author:
+            deprecated("[config file] coloring 'authors' has been renamed 'author', however a " +
+                       "coloring for 'author' already exists. 'authors' will be ignored")
```
```suggestion
            deprecated("[config file] ignoring deprecated coloring 'authors' because a " +
                       "coloring for 'author' already exists.")
```

jameshadfield replied:
Thanks! Fixed up

genehack commented on augur/util_support/auspice_config.py:
```diff
@@ -0,0 +1,234 @@
+    merged = configs[0]
+    for overlay in configs[1:]:
+        # We could leverage the config schemas here in the future if desired
+        merged = _merge_scalar(merged, overlay, 'title')
+        merged = _merge_lists(merged, overlay, 'colorings', lambda x,y: x.get('key')==y.get('key'))
+        merged = _merge_lists(merged, overlay, 'geo_resolutions', lambda x,y: _geo_resolution_id(x)==_geo_resolution_id(y))
+        merged = _merge_lists(merged, overlay, 'maintainers', lambda x,y: x.get('name')==y.get('name'))
+        merged = _merge_scalar(merged, overlay, 'build_url')
+        merged = _merge_scalar(merged, overlay, 'build_avatar')
+        merged = _merge_lists(merged, overlay, 'filters', lambda x,y: x==y)
+        merged = _merge_dicts(merged, overlay, 'display_defaults')
+        merged = _merge_lists(merged, overlay, 'panels', lambda x,y: x==y)
+        merged = _merge_lists(merged, overlay, 'data_provenance', lambda x,y: x.get('name')==y.get('name'))
+        merged = _merge_lists(merged, overlay, 'metadata_columns', lambda x,y: x==y)
```
Suggestion: convert this block into a data structure of the important bits (the merging function, the key name, and the identity lambda) into a data structure, and then loop over that to do the work, a la the `DEPECATIONS` data structure above.

jameshadfield replied:
I didn't get to this - maybe I will, I'm not certain. I agree the current form is overly repetitive, but it's very self explanatory as well!

genehack replied:
Yeah, fair enough — I will admit to a fondness (weakness?) for the "loop a data structure" pattern that was used for `DEPRECATIONS` but this was very much a non-blocking suggestion.

joverlee521 replied:
Hmm, I can see the merge not behaving as expected if we add a new key to the auspice config but forget to add to the merge function here. I think a loop data structure could help flag any keys that were missed in the merge.

genehack commented:
> We could additionally validate the merged config if people want to, although the current implementation guarantees that valid configs produce a valid merged config.

I think re-validating the merged config is a good idea — even if it's the equivalent of a "this never happens" `else` clause in a conditional — because if somehow an invalid config _was_ produced, I have to imagine it would be _extremely_ confusing to debug.

j23414 commented:
Only commenting from a user perspective, not coding:

+1 For this feature, because I'd like to use it to simplify [dengue's set of auspice configs](https://github.com/nextstrain/dengue/blob/0d10460107c96171ef5261bb7a648d5e29a6edad/phylogenetic/rules/export.smk#L40-L148)

> merging of lists is done by extending the original list, rather than simply replacing it.

Cool and on-board with this! And I imagine from a user's perspective, a usage case would be similar to 

```
augur export --auspice-config auspice_config_base.json  add-serotype-specific-coloring-and-configs.json ...
augur export --auspice-config auspice_config_base.json  add-genotype-specific-coloring-and-configs.json ...
```

jameshadfield commented on augur/utils.py:
```diff
@@ -435,6 +435,8 @@ def _read_genbank(reference, feature_names):
     return features
 
+# NOTE `read_config` is only used by `export_v1.py` and should be removed when that
+# file is removed.
```
Good catch - I thought the code was going to diverge but it never did. `export_v1.py` now has
```py
from .util_support.auspice_config import read_json as read_config
```
so when we eventually remove export v1 in #1754 this should be very clean.

jameshadfield commented:
Force pushed to address requested changes (see comments above) and rebased onto latest master.

> I think re-validating the merged config is a good idea

Done, and also added a `--output-auspice-config` argument as discussed in dev chat earlier today

Read The Docs is [failing](https://app.readthedocs.org/projects/nextstrain-augur/builds/27663422/) as it failed to check out the branch, which I presume is temporary...

joverlee521 commented on augur/util_support/warnings.py:
```diff
@@ -0,0 +1,29 @@
+def warn(message):
+    warnings.warn(message, UserWarning, stacklevel=2)
+
+def configure_warnings():
+    # we must only set these when someone runs `augur export v2` (i.e. run_v2() is called)
+    # else they will apply to _all_ augur commands due to the way all commands are pulled
+    # in by the augur runner (augur/__init__.py)
+    def customformatwarning(message, category, filename, lineno, line=None):
+        if category.__name__ == "UserWarning":
+            return "WARNING: {}\n\n".format(message)
+        if category.__name__ == "DeprecationWarning":
+            return "DEPRECATED: {}\n\n".format(message)
+        return "{}\n".format(message)
+
+    warnings.formatwarning = customformatwarning
+    warnings.simplefilter("default") # show DeprecationWarnings by default
```
> Over time we will hopefully make use of these functions in other parts of Augur.

Nice! So eventually the warnings would be centrally configured and then we'd replace a lot of the `print_err("WARNING...)` calls with `warn()`.

jameshadfield replied:
Yup! Looking at this after a week I should update the comment to not imply that this can only be used for export v2

jameshadfield replied:
[comment updated](https://github.com/nextstrain/augur/compare/23595ee4deb4eebf66d73272f20d54dbfc420f03..b294cbcdb89cef2c2ea4ae7e5a3f8957a67abb5d)

joverlee521 commented on augur/export_v2.py:
```diff
@@ -899,10 +888,10 @@ def register_parser(parent_subparsers):
             Supplying both is fine too, command line args will overrule what is set in the config file!"
     )
-    config.add_argument('--auspice-config', metavar="JSON", help="Auspice configuration file")
+    config.add_argument('--auspice-config', metavar="JSON", nargs='+', action=ExtendOverwriteDefault, default=[], help="Auspice configuration file(s)")
```
I think it'd be helpful to explicitly state in the help message that if multiple config files are provided then they would be merged and give a brief description of how the values are merged.

jameshadfield replied:
I don't think I can convey _how_ the merging behaviour works within the `--help` message, but I can totally say they'll be merged. Separately I plan to write a docs page, but I was going to leave it until it's been in use for a while internally so that it's easier to make breaking changes, but I can add them now if you think that's preferable?

jameshadfield replied:
[Updated the argument group's help message](https://github.com/nextstrain/augur/compare/23595ee4deb4eebf66d73272f20d54dbfc420f03..b294cbcdb89cef2c2ea4ae7e5a3f8957a67abb5d)

joverlee521 replied:
The updated help message is fine for now! I just know there will eventually be questions on how the auspice configs are merged once this is more widely used.

joverlee521 reviewed (changes requested):
The general merge behavior makes sense to me! My only blocking comment is to add some documentation to the help message to clarify how multiple Auspice configs are merged.

jameshadfield commented:
We now have new CI failures after [this force-push](https://github.com/nextstrain/augur/compare/23595ee4deb4eebf66d73272f20d54dbfc420f03..b294cbcdb89cef2c2ea4ae7e5a3f8957a67abb5d), however I cannot reproduce them locally. Diffing the output from locally run tests & the [CI logs](https://github.com/nextstrain/augur/actions/runs/14234325053/job/39890823133) implies it might be the 'warning' below, which would be completely unrelated to the force push 🫠 

```
=================================== FAILURES ===================================
_________________________________ test_flake8 __________________________________
tests/test_flake8.py:9: in test_flake8
    assert result.returncode == 0, "flake8 exited with errors"
E   AssertionError: flake8 exited with errors
E   assert 1 == 0
E    +  where 1 = CompletedProcess(args=['flake8'], returncode=1).returncode
        result     = CompletedProcess(args=['flake8'], returncode=1)
=============================== warnings summary ===============================
augur/export_v2.py:949
  /home/runner/work/augur/augur/augur/export_v2.py:949: DeprecationWarning: Nesting mutually exclusive groups is deprecated.
    root_sequence = root_sequence_group.add_mutually_exclusive_group()
```
